crates/lang/src/reflect/trait_def/registry.rs

```diff
@@ -61,3 +67,37 @@ where
 {
     type Env = E;
 }
+
+impl<E> ToAccountId<E> for TraitDefinitionRegistry<E>
+where
+    E: ink_env::Environment,
+{
+    /// `to_account_id` is not allowed for `TraitDefinitionRegistry`.
+    ///
+    /// We insert markers for these errors in the generated contract code.
+    /// This is necessary since we can't check these errors at compile time
+    /// of the contract.
+    /// `cargo-contract` checks the contract code for these error markers
+    /// when building a contract and fails if it finds markers.
+    #[cold]
+    fn to_account_id(&self) -> E::AccountId {
+        /// We enforce linking errors in case this is ever actually called.
+        /// These linker errors are properly resolved by the cargo-contract tool.
+        extern "C" {
+            fn __ink_enforce_error_to_account_id() -> !;
+        }
+        unsafe { __ink_enforce_error_to_account_id() }
+    }
```

Comment:
So I'm still trying to understand why this is needed.

Previously these were only in the actual method code, not the `ToAccountId` implementation. E.g, `<TraitDefinitionRegistry as $trait>::$trait_method()`. This was to uphold the [following guarantee from the trait](https://github.com/paritytech/ink/blob/1188ee71284dc1e6b315311561f916fabd668b09/crates/lang/src/reflect/trait_def/registry.rs#L27):

> ///   `TraitDefinitionRegistry` with stub implementations for all methods that
> ///   guarantee that they are never called.

Looking at your changes it looks like we're now going through a `...to_account_id().fire()` sequence in the trait method implementations, so is this here as a way to continue to uphold that?

Reply:
Previously we generated implementation of trait for `TraitDefinitionRegistry` only to store `__ink_TraitInfo`. It is why methods were stubbed with that error. The idea is that no one calls these methods.

In the current change that implementation is inherited automatically for `TraitDefinitionRegistry` because we added the blanket implementation. But methods are not empty(stubbed), those contain code for cross-contract calls that are using `to_account_id`.

I added an error into `to_account_id` to prevent the calling of methods from `TraitDefinitionRegistry`. 
`<TraitDefinitionRegistry as $trait>::$trait_method()` is using `...to_account_id().fire()` inside and it will cause compilation error like before.

Reply:
Okay yeah, we're on the same page then.

Reply:
Can you also update this doc comment on the `TraitDefinitionRegistry` struct?

> /// - Furthermore the ink! codegen implements the ink! trait definition for the
> ///   `TraitDefinitionRegistry` with stub implementations for all methods that
> /// guarantee that they are never called.

Reply:
I think that comment is still true because the implementation of these methods is really stubbed. Calling these methods will cause a compilation error.

Reply:
Maybe it's pedantic, but they're not stubs in my opinion. The only stub is `to_account_id()`